File: LearnOpencv/7/python/main.py
import cv2
import numpy as np

# 读取彩色图像
image = cv2.imread("C:/github/lib/picture/p2.jpg")
if image is None:
    print("Error: Could not open or find the image!")
    exit()

# 调整大小便于显示    
image=cv2.resize(image,(320,320),interpolation=cv2.INTER_AREA)

# 通道分离
hsv_image=cv2.cvtColor(image,cv2.COLOR_BGR2HSV)

# 定义红色的HSV范围
lower_red1=np.array([0,100,100])
upper_red1=np.array([10,255,255])  
lower_red2=np.array([160,100,100]) 
upper_red2=np.array([180,255,255])

# 获得红色区域掩码
mask_red1=cv2.inRange(hsv_image,lower_red1,upper_red1)
mask_red2=cv2.inRange(hsv_image,lower_red2,upper_red2)
mask_red=cv2.bitwise_or(mask_red1,mask_red2)

# 定义蓝色的HSV范围
lower_blue=np.array([100,100,100])   
upper_blue=np.array([140,255,255])

# 获得蓝色区域掩码
mask_blue=cv2.inRange(hsv_image,lower_blue,upper_blue)

# 利用掩码分别提取红色与蓝色部分
red_extracted=cv2.bitwise_and(image,image,mask=mask_red)
blue_extracted=cv2.bitwise_and(image,image,mask=mask_blue)

# 显示图像
cv2.imshow("Original Image",image)
cv2.imshow("Red_Extraced",red_extracted)
cv2.imshow("Blue_Extraced",blue_extracted)

# 等待按键关闭
cv2.waitKey(0)

File: LearnOpencv/7/c++/main.cpp
#include <iostream>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/opencv.hpp>
using namespace std;
using namespace cv;

int main()
{
    //读取图像
    Mat img=imread("C:/github/lib/picture/p2.jpg");
    
    //通道分离
    Mat hsv_img;
    cvtColor(img,hsv_img,COLOR_BGR2HSV);
    
    //定义红色的HSV范围
    Scalar lower_red1(0, 100, 100);
    Scalar upper_red1(10, 255, 255);
    Scalar lower_red2(160, 100, 100);
    Scalar upper_red2(180, 255, 255);

    //获得红色区域掩码
    Mat mark_red1,mark_red2,mark_red;
    inRange(hsv_img,lower_red1,upper_red1,mark_red1);
    inRange(hsv_img,lower_red2,upper_red2,mark_red2);
    bitwise_or(mark_red1,mark_red2,mark_red);

    //定义蓝色的HSV范围
    Scalar lower_blue(100, 100, 100);
    Scalar upper_blue(140, 255, 255);

    //获得蓝色区域掩码
    Mat mark_blue;
    inRange(hsv_img,lower_blue,upper_blue,mark_blue);

    //利用掩码分别提取红色与蓝色部分
    Mat red_extracted,blue_extracted;
    bitwise_and(img,img,red_extracted,mark_red);
    bitwise_and(img,img,blue_extracted,mark_blue);

    //显示图像
    imshow("Original Img",img);
    imshow("Red_Extracted",red_extracted);
    imshow("Blue_Extracted",blue_extracted);

    //等待按键关闭
    waitKey(0);
    return 0;
}